### PR TITLE
18749 - add amalgamation statement to review

### PR DESCRIPTION
**Pull request by tshyun24**
*Issue #:* /bcgov/entity#18749

*Description of changes:*

1. add Amalgamation Statement component (Jan5 updated)
image

2. validation getter (Jan5 updated)
if nothing selected and click `File and Pay`
image
once either of the option being selected
image

3. Able to save and restore

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the bcrs-entities-create-ui license (Apache 2.0).

### CODE REVIEW

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-create-dev--pr-609-ft6cgzmy.web.app

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-create-dev--pr-609-ft6cgzmy.web.app

File: src/interfaces/store-interfaces/state-interfaces/amalgamation-state-interface.ts
Comment (tshyun24) [inline src/interfaces/store-interfaces/state-interfaces/amalgamation-state-interface.ts:1]: Add this for validation

File: src/mixins/filing-template-mixin.ts
Comment (tshyun24) [inline src/mixins/filing-template-mixin.ts:1]: Change it to `!=null`, otherwise the file cannot be saved when the user selected not court approval option

**Comment by severinbeauvais** [comment]
Re:
> 1. add Amalgamation Statement component
>    image

The label colour is incorrect. If I inspect the sample, the color is `#0009`. Instead, it should be $gray7. See the specifications in the uxpin: 

image

 (The text colour is correct.)

**Comment by severinbeauvais** [comment]
Re:
> 2. validation getter
>    if nothing selected and click `File and Pay`
>    image

The red bar should extend all the way up.

File: src/mixins/filing-template-mixin.ts
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:1]: OK, but it should be `!== null`.

(Google it if you don't know why.)

File: src/store/store.ts
Comment (severinbeauvais) [inline src/store/store.ts:594]: A temporary variable makes sense for lines 592 and 593 since they do some computation and it looks cleaner to not have that on lines 601 and 602.

In your case, I recommend just putting the getter on line 603.

File: src/views/AmalgamationRegular/ReviewConfirm.vue
Comment (severinbeauvais) [inline src/views/AmalgamationRegular/ReviewConfirm.vue:181]: You don't need the v-card here. It was a placeholder for your new component.

Right now you have a v-card inside a v-card, which is causing some spacing and error validation issues. (It's not terrible to have a v-card inside a v-card, but it's not needed here. The other components on this page don't do it.)

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [inline src/components/Amalgamation/AmalgamationStatement.vue:116]: You also need to set the label colour to $gray9.

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [inline src/components/Amalgamation/AmalgamationStatement.vue:106]: It's often better to put the layout properties right in the template (using the Vuetify helper classes), so that a future developer doesn't have to look up classnames/selectors to determine what properties are being applied, and to facilitate layout updates. (Don't you hate having to look up selectors in the CSS to see where various properties come from? This is what we are trying to avoid if possible.)

The other properties here (probably) don't have helper classes, so they're fine here. But it would be even better -- future state -- if we had base/common classes for things like radio labels, statements, headers, etc. Then we wouldn't need any CSS here; we would just label everything according to their classname. This is the "ideal case". For now, properties here are OK. But... if you find an opportunity to make base/common classes for things, please go ahead :)

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [inline src/components/Amalgamation/AmalgamationStatement.vue:92]: FYI, there's a shortcut syntax for this:
```
  // Emit a boolean (validation) which is either option being selected
  @Emit('valid')
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  private amalgamationStatementValid (valid: boolean): void {}
```

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [inline src/components/Amalgamation/AmalgamationStatement.vue:84]: Remove debugging before merging PR.

After removing this line, you could use the shortcut syntax here, too.

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [inline src/components/Amalgamation/AmalgamationStatement.vue:81]: If you use JSDoc syntax for the comment, you can have nice contextual help, like this:

image

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [inline src/components/Amalgamation/AmalgamationStatement.vue:72]: This is a clever way to specify the type but I'm not sure if I love it. Did you see somewhere that it is a best practise?

The issue I have with this is that you need to know the property name in the interface, but I don't know if that's worse than just saying it's a boolean (which you need to know in this file anyway).

Another thing I'm not sure I like is the syntax, which looks like a quoted object property (compared to a named object property). There are lint rules (strict-property-naming) to enforce the named style, and lots to read on the subject... My conclusion is that quoted properties are to be avoided,

(This is not the same as having object properties in quotes. That's another topic.)

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [inline src/components/Amalgamation/AmalgamationStatement.vue:78]: 1. Why do you need the conditional on line 76? If the getter value is null then you could simply reassign null to the local variable.

2. Always use `===` or `!==` in comparisons.

3. This fails if the component is mounted before the getter has a valid value (ie, the draft filing has been loaded). Fortunately, I think App.vue is written so that the pages aren't displayed (mounted) until the draft filing has been loaded, so it works here, but don't count on it in other places. I think we don't have a consistent method of having local variables synced with props or store variables (which are both read-only), so this is always a bit tricky. In other code, we have a watcher that sets the local initially ("immediate: true") and whenever the prop/getter changes. I think that's safer code but I think it's still overly complex. If you think of a nicer way to do this, let me know.

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [inline src/components/Amalgamation/AmalgamationStatement.vue:31]: When possible, you should use the best tag/element for its content -- in this case, it would be better to use the paragraph element instead of the generic div element.

See also https://www.w3schools.com/html/html5_semantic_elements.asp

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [inline src/components/Amalgamation/AmalgamationStatement.vue:108]: We have CSS variables for font sizes. Please use them. (It allows for easier global searching for where different sizes are used.)

**Review by severinbeauvais — COMMENTED** [review]
Could you please add a unit test file for your new component? Thanks.

File: src/mixins/filing-template-mixin.ts
Comment (JazzarKarim) [inline src/mixins/filing-template-mixin.ts:265]: I don't think you need to change this as the previous syntax does exactly this!

Comment (tshyun24) [reply src/mixins/filing-template-mixin.ts:265]: could you plz explain more? cause I try without this change, when I select "without court approval", the selection cannot be saved

Comment (JazzarKarim) [reply src/mixins/filing-template-mixin.ts:265]: OK, it's fine. But please if you can change it to `!==`. Thanks!

Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:265]: I think `courtApproval` may be True or False, which is why the check for null is needed.

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (tshyun24) [reply src/components/Amalgamation/AmalgamationStatement.vue:78]: > 1. Why do you need the conditional on line 76? If the getter value is null then you could simply reassign null to the local variable.

yeahhh, I just realize that.

> 3.

I think I will keep the same with other file, using ("immediate: true"). But I am wondering what does "other places" in ` I think App.vue is written so that the pages aren't displayed (mounted) until the draft filing has been loaded, so it works here, but don't count on it in other places. ` means?

Comment (severinbeauvais) [reply src/components/Amalgamation/AmalgamationStatement.vue:78]: > But I am wondering what does "other places" in (snip) means?

In this web app, I think it is safe to assume that most components aren't created/mounted until the data is loaded, but in a different UI this may not be true. Maybe Filings UI, maybe a future web app you work on, etc :)

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (tshyun24) [reply src/components/Amalgamation/AmalgamationStatement.vue:72]: > Did you see somewhere that it is a best practise?

Nope, I just go through other files and find that it makes me more clear to write down the property name at here.

> but I don't know if that's worse than just saying it's a boolean

I am thinking should I simply write `courtApproval = null as boolean` or do `courtApproval: AmalgamationStateIF['courtApproval'] = null as boolean`? The first one is more simple, the second one may contain more information but don't know how to avoid the quoted properties

Comment (severinbeauvais) [reply src/components/Amalgamation/AmalgamationStatement.vue:72]: Yes, exactly.

Let's leave your current code for now, use it a bit, and decide later. It's easy to change later if we decide against it.

Comment (severinbeauvais) [reply src/components/Amalgamation/AmalgamationStatement.vue:72]: Shaoyun, you can resolve this conversation after you've read this.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-create-dev--pr-609-ft6cgzmy.web.app

File: src/mixins/filing-template-mixin.ts
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:1]: Looks good.

File: src/store/store.ts
Comment (severinbeauvais) [reply src/store/store.ts]: Looks good.

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [reply src/components/Amalgamation/AmalgamationStatement.vue:122]: Looks good.

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [reply src/components/Amalgamation/AmalgamationStatement.vue]: Looks good.

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [reply src/components/Amalgamation/AmalgamationStatement.vue]: Looks good.

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-create-dev--pr-609-ft6cgzmy.web.app

**Comment by bcregistry-sre** [comment]
Temporary Url for review: https://business-create-dev--pr-609-ft6cgzmy.web.app

File: tests/unit/AmalgamationStatement.spec.ts
Comment (severinbeauvais) [inline tests/unit/AmalgamationStatement.spec.ts:1]: Please fix GitHub Actions linting issue identified below. Thanks!

If you are using VS Code, you may have to set the default indentation (to 2) for a new file in its settings.

File: src/mixins/filing-template-mixin.ts
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts:1]: I'm resolving this conversation because Karim started another conversation to discuss whether or not this is needed (which it is).

File: tests/unit/AmalgamationStatement.spec.ts
Comment (tshyun24) [reply tests/unit/AmalgamationStatement.spec.ts:1]: Hii Sev, please see the newest comment, which should fix the linting issue -)

File: src/mixins/filing-template-mixin.ts
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts]: Shaoyun, since this caused confusion once, and to avoid a future developer trying to "optimize" this and actually break it, please add a code comment explaining why you check for not null (ie, save value if it's True or False).

Otherwise, this code looks good to me.

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [reply src/components/Amalgamation/AmalgamationStatement.vue]: Please use CSS variable `$px-16` for this font size (and also the other font sizes in this file). Thx.

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [reply src/components/Amalgamation/AmalgamationStatement.vue]: Changed to "p". Looks good. Resolving.

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [reply src/components/Amalgamation/AmalgamationStatement.vue]: How about changing your function comments from `// Emit a ...` to
```
/**
  * Emits a ...
  * @param courtApproval The value to emit.
  */
```

(For all 3 functions with non-JSDoc syntax.)

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [reply src/components/Amalgamation/AmalgamationStatement.vue]: Thanks for the change. Looks good.

File: src/views/AmalgamationRegular/ReviewConfirm.vue
Comment (severinbeauvais) [reply src/views/AmalgamationRegular/ReviewConfirm.vue]: Looks good.

**Review by severinbeauvais — APPROVED** [review]
Just a couple small things otherwise I approve this.

File: tests/unit/AmalgamationStatement.spec.ts
Comment (severinbeauvais) [reply tests/unit/AmalgamationStatement.spec.ts:1]: Yes, you fixed the indentation in the latest commit. Thanks! Resolving.

**Comment by tshyun24** [comment]
> Just a couple small things otherwise I approve this.

Thank you so much to your patient Sev!!! Lots of useful reviews🙏

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [inline src/components/Amalgamation/AmalgamationStatement.vue:94]: Delete

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (severinbeauvais) [inline src/components/Amalgamation/AmalgamationStatement.vue:93]: Also update comment for setCourtApproval ?

File: src/mixins/filing-template-mixin.ts
Comment (severinbeauvais) [reply src/mixins/filing-template-mixin.ts]: The updated comment is fine by me. Resolving.

File: src/components/Amalgamation/AmalgamationStatement.vue
Comment (tshyun24) [reply src/components/Amalgamation/AmalgamationStatement.vue:93]: Just wondering because it's does not have `param` and `return`, should I also update the comment to JSDoc syntax or leave it now?

Comment (severinbeauvais) [reply src/components/Amalgamation/AmalgamationStatement.vue:93]: Sure. That way you'll get a nice hover help panel for the function :)

**Comment by severinbeauvais** [comment]
@ketaki-deodhar If you have a moment today, please review and merge this. If not, just merge it. Please and thanks!

**Comment by ketaki-deodhar** [comment]
> @ketaki-deodhar If you have a moment today, please review and merge this. If not, just merge it. Please and thanks!

Looking at it now Sev!

**Review by ketaki-deodhar — APPROVED** [review]
Looks good!

**Pull request merged**
